pages/clubs.jsx

```diff
@@ -82,63 +84,176 @@ var BottomCTA = React.createClass({
+    var clubState = _.pick(this.state,
       'name', 'website', 'description', 'location'
-    ), function(err) {
-      if (err) {
-        window.alert("Alas, an error occurred. Please try again later!");
-        console.log(err);
-      } else {
-        window.alert("Your club has been added!");
-        this.hideModal();
-      }
-    }.bind(this));
+    );
+
+    e.preventDefault();
+    this.setState({
+      step: this.STEP_WAIT_FOR_NETWORK,
+      networkError: false
+    });
+    if (this.props.club) {
+      clubState = _.extend({}, this.props.club, clubState);
+
+      // For now, let's always force re-geocoding.
+      clubState.latitude = null;
+      clubState.longitude = null;
+
+      teachAPI.changeClub(clubState, this.handleNetworkResult);
+    } else {
+      teachAPI.addClub(clubState, this.handleNetworkResult);
+    }
+  },
+  handleNetworkResult: function(err, data) {
+    if (!this.isMounted()) {
+      return;
+    }
+    this.setState({
+      networkError: !!err,
+      step: err ? this.STEP_FORM : this.STEP_SHOW_RESULT,
+      result: err ? null : data
+    });
+  },
+  handleSuccessClick: function() {
+    this.hideModal();
+    this.props.onSuccess(this.state.result);
+  },
+  handleJoinClick: function() {
+    this.hideModal();
+    this.transitionTo('join');
   },
   render: function() {
+    var content, isFormDisabled;
+    var isAdd = !this.props.club;
+    var action = isAdd ? "add" : "change";
+    var modalTitle = isAdd ? "Add Your Club To The Map"
+                           : "Change Your Club";
+
+    if (this.state.step == this.STEP_AUTH) {
+      content = (
+        <div>
+          <p>Before you can {action} your club, you need to log in.</p>
+          <button className="btn btn-primary btn-block"
+           onClick={this.getTeachAPI().startLogin}>Log In</button>
+          <button className="btn btn-default btn-block"
+           onClick={this.handleJoinClick}>Create an account</button>
+        </div>
+      );
+    } else if (this.state.step == this.STEP_FORM ||
+               this.state.step == this.STEP_WAIT_FOR_NETWORK) {
+      isFormDisabled = (this.state.step == this.STEP_WAIT_FOR_NETWORK);
+      content = (
+        <div>
+          {this.state.networkError
+           ? <div className="alert alert-danger">
+               <p>Unfortunately, an error occurred when trying to {action} your club.</p>
+               <p>Please try again later.</p>
+             </div>
+           : null}
+          <form onSubmit={this.handleSubmit}>
+            <fieldset>
+              <label>What is the name of your Club?</label>
+              <input type="text" placeholder="We love creative Club names"
+               disabled={isFormDisabled}
+               required
+               valueLink={this.linkState('name')} />
+            </fieldset>
+            <fieldset>
+              <label>Where does it take place?</label>
+              <input type="text" placeholder="Type in a city or a country"
```

Comment:
can we add validation to these field to not only make it `required` but also making sure is not a `space` character? I just enter ` ` and it worked, but this seems to be weird to accept such input though is not an error.